### PR TITLE
Unifying names of "smart" combinators + adding combinators in CArray

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** architecture

This is part of "improving" the API where "improving" means here easier to learn and remember, more regular, even if it implies extra "burden" for who already uses the API.

Concretely:
- We model the name of `smartmap` and `smartfoldmap` on the name of the underlying combinators (and deprecate other names).
- We add two new smart combinators that will be useful for coq/ceps#34.

More generally, here are questions to efficiency experts:
- Couldn't we have a more precise name than `smart` to describe these combinators? For instance, I'm thinking at `homogeneous` or, shortlier, `homo`.
- What about having a specific name space, say `List.Homo`, `Array.Homo` for such combinators, as we have e.g. `Array.Fun1`, reusing then the "standard" naming scheme within the given name spaces?
- What about using more pervasively these combinators in the implementation whenever the `map` is homogeneous?

I can eventually do all these changes.

PS: This is part of ongoing implementation on coq/ceps#34, isolating these commits for easier "domain-specific" reviewing and discussion.

### CODE REVIEW

File: clib/cArray.mli
Comment (ejgallego) [inline clib/cArray.mli]: This should be `[@@ocaml.deprecated "Please use [smart_fold_left_map]"]` Ditto for the rest of definitions.

**Review by ejgallego — COMMENTED** [review]
I have no opinion on the renaming, in principle I am not sure it is worth the cost, but I am OK with this if other people agree.

Need to review the patch more in depth but it looks ok, sauf for the lack of deprecate attribute [this should be fixed]

**Review by ppedrot — DISMISSED** [review]
I'd support the move to a `Smart` submodule of `Array` rather than aimlessly renaming the function once just to rename it immediately if we decide to modulify it altogether...

**Comment by ejgallego** [comment]
> I'd support the move to a Smart submodule of Array rather than aimlessly renaming the function once just to rename it immediately if we decide to modulify it altogether...

Umm, I'd need to think more about it, not sure if `let open Array.Smart in map ...` is gonna be too risky. What are the advantages of adding the `Smart` module?

File: clib/cArray.mli
Comment (herbelin) [reply clib/cArray.mli]: Done.

**Comment by herbelin** [comment]
Both `List.Fun1.map` and `List.Smart.map` (or whatever names we give to them) should actually be available under the name `List.map` from the point of view of a user of the `List` module, if ever OCaml had the possibility to write ad hoc polymorphic functions.

**Comment by herbelin** [comment]
> Umm, I'd need to think more about it, not sure if `let open Array.Smart in map ...` is gonna be too risky. 

We can also do `let open Array in Smart.map ...`.

> What are the advantages of adding the Smart module?

To ensure a systematic naming for optimized versions of combinators.

On the other side, this would mean moving the `smart` functions possibly far from the heterogeneous ones (or to extend the `Smart` module one by one, but the OCaml syntax for doing so is not very encouraging).

In any case, I don't think that this PR needs special postponent. Let's have the discussion now, and, if nothing special to say now, new steps can be done later on.

@ppedrot: ok to be the assignee?

**Comment by ejgallego** [comment]
I like the `let open Array in Smart.map`, nice!

**Review by maximedenes — DISMISSED** [review]
I think I'm missing the point of the `smartmap` -> `smart_map` renaming (and the others). But I agree with @ppedrot, if the motivation is uniformity, we should go directly for a `Smart` module, to reduce to noise for developers.

**Comment by herbelin** [comment]
> I think I'm missing the point of the smartmap -> smart_map renaming (and the others).

This was mostly to see `smart` as an individual brick for composing names. The same way as we have `fold_map` and not `foldmap`, I found more consistent to have `smart_filter` and `smart_map`. But I agree that we could also see `smart` as "agglutinative".

So, it seems to converge to try to have a Smart module?

**Comment by herbelin** [comment]
I started to experiment with a `Smart` module but this is not pure gain. In particular, this forces an organization of the API where the optimized variants are isolated from the rest of the API. Before continuing processing, I'd like to be sure that we are ready to have the specialized and optimized `Smart` module coming well after the most general variants of the same kind (which does not prevent using cross-references in the comments though).

So, at the current time, I got satisfied with just keeping a `smart_` prefix.

**Review by PierreCorbineau — APPROVED** [review]

**Review by barras — APPROVED** [review]

**Comment by ppedrot** [comment]
> So, at the current time, I got satisfied with just keeping a smart_ prefix.

But then it doesn't bring much. And I'm definitely not in favour of a cosmetic change that affects many lines of code for no obvious advantage...

**Comment by herbelin** [comment]
> But then it doesn't bring much. And I'm definitely not in favour of a cosmetic change that affects many lines of code for no obvious advantage...

Another point: `Smart` looks to me a bit uninformative. While we are at moving to a module, shouldn't we rather use a name which emphasizes that these are specialized version of the previous version for a return type which is the same as one of the input type?

Note also that the renaming done in this PR actually goes further than its title: the `smart` part can also be considered in conjunction with #7249.

**Review by forestjulien — APPROVED** [review]

**Comment by herbelin** [comment]
Update:
- I made a module `Smart` for `Array`, `List`, `Option`, `Rtree`, `Map`. So, all of these have a uniform naming scheme regarding the (few) smart functions they provide
- In passing, I inlined `Fun1` so that it can be accessed with `Array.Fun1` rather than `CArray.Fun1`.
- I renamed the PR so that it reflects that it is also about `List`, `Option`, `Rtree`, `Map`.
- I (hopefully) did not miss a compatibility ML attribute

@maximedenes: I dismissed your review as it seems I answered your two questions:
1. it is not only about `smartmap`, it is also e.g. about `smartfoldmap` whose name is not "standard"
2. I made a `Smart` module

Before final merge, it might be better to eliminate the intermediate step `smart_`.

More generally, I'm pretty optimistic about the current dynamic around the API question and about finding a path through the dialectic between stability vs easy-to-explain code organization. A typical question: will we try to see `constr.ml` and `eConstr.ml` under the same "library with methods over a type" glasses as we see `Array` or `List`, and document it along lines consistent with what we do for `Array` or `List`. For instance, @ppedrot, you initiated `constr.ml`, but it has not been clear to me yet whether the intent was to go towards a library view at `Constr`. What is your opinion? For instance, do you adhere to the view of merging `Constr` and `Term` and to dispatch `Termops` into `Constr` or `EConstr`, as discussed these days? (Sorry if you already expressed an opinion that I missed or forgot.)

**Comment by ppedrot** [comment]
> Before final merge, it might be better to eliminate the intermediate step smart_.

@herbelin I'm now fine with the final code, but indeed squashing the PR is welcome.

> For instance, @ppedrot, you initiated constr.ml, but it has not been clear to me yet whether the intent was to go towards a library view at Constr. What is your opinion? For instance, do you adhere to the view of merging Constr and Term and to dispatch Termops into Constr or EConstr, as discussed these days? (Sorry if you already expressed an opinion that I missed or forgot.)

I don't know, I have kept changing my mind on that topic. Originally I think that the `Constr` vs. `Term` distinction was to emphasize that some combinators need to know the internals to be implemented efficiently, and stuff in `Term` are derived using these low-level combinators. But I'm probably more leaning towards the merge of `Constr` and `Term` as of today.

**Review by mattam82 — APPROVED** [review]

**Comment by ppedrot** [comment]
Is everybody happy with the current changes? If so, @herbelin can you rebase and squash so that I can merge the PR?

**Comment by ejgallego** [comment]
> Is everybody happy with the current changes?

I am OK.

**Comment by herbelin** [comment]
I squashed and dispatched changes module by module (but maybe it is overkill after all).

**Pull request merged**
